packages/blade/src/components/Input/TextInput/TextInput.tsx

```diff
@@ -0,0 +1,227 @@
+  onChange,
+  onBlur,
+  isDisabled,
+  necessityIndicator,
+  validationState,
+  errorText,
+  helpText,
+  successText,
+  isRequired,
+  icon,
+  prefix,
+  showClearButton,
+  onClearButtonClick,
+  isLoading,
+  suffix,
+  autoFocus,
+  keyboardReturnKeyType,
+  autoCompleteSuggestionType,
+}: TextInputProps): ReactElement => {
+  const textInputRef = React.useRef<HTMLInputElement | TextInputReactNative>(null);
+  const [shouldShowClearButton, setShouldShowClearButton] = useState(false);
+
+  const renderInteractionElement = (): ReactNode => {
+    if (isLoading) {
+      return <InfoIcon size="medium" color="surface.text.subtle.lowContrast" />;
+    }
+
+    if (shouldShowClearButton) {
+      return (
+        <IconButton
+          size="medium"
+          icon={CloseIcon}
+          onClick={() => {
+            if (isEmpty(value) && textInputRef.current) {
+              // when the input field is uncontrolled take the ref and clear the input and then call the onClearButtonClick function
+              if (isReactNative(textInputRef.current)) {
+                textInputRef.current.clear();
+              } else if (textInputRef.current instanceof HTMLInputElement) {
+                textInputRef.current.value = '';
+              }
+            }
+            // if the input field is controlled just call the click handler and the value change shall be left upto the consumer
+            onClearButtonClick?.();
+            setShouldShowClearButton(false);
+          }}
+          accessibilityLabel="Clear Input Content"
+        />
+      );
+    }
+
+    return null;
+  };
+
+  return (
+    <BaseInput
+      id="textinput"
+      ref={textInputRef as React.Ref<HTMLInputElement>}
+      label={label}
+      labelPosition={labelPosition}
+      placeholder={placeholder}
+      defaultValue={defaultValue}
+      value={value}
+      name={name}
+      maxCharacters={maxCharacters}
+      onChange={({ name, value }) => {
+        if (showClearButton && value?.length) {
+          // show the clear button when the user starts typing in
+          setShouldShowClearButton(true);
+        }
+
+        if (shouldShowClearButton && !value?.length) {
+          // hide the clear button when the input field is empty
+          setShouldShowClearButton(false);
+        }
+
+        onChange?.({ name, value });
+      }}
+      onBlur={onBlur}
+      isDisabled={isDisabled}
+      necessityIndicator={necessityIndicator}
+      isRequired={isRequired}
+      leadingIcon={icon}
+      prefix={prefix}
+      interactionElement={renderInteractionElement()}
+      suffix={suffix}
+      validationState={validationState}
+      errorText={errorText}
+      helpText={helpText}
+      successText={successText}
+      trailingFooterSlot={(value) => {
+        return maxCharacters ? (
+          <Box marginTop="spacing.2" marginRight="spacing.1">
+            <CharacterCounter currentCount={value?.length ?? 0} maxCount={maxCharacters} />
```

Comment:
accessibility good to have feature:

Announce how many characters are left, Right now SR users have no idea.

Let's track this on an issue, we can do it as an enhancement.

Reply:
good catch, completely missed that